api/saml/controller.py

```diff
@@ -0,0 +1,249 @@
+        )
+        params = {
+            self.ERROR: problem_detail_json
+        }
+        redirect_uri = self._add_params_to_url(redirect_uri, params)
+
+        return redirect_uri
+
+    def _get_redirect_uri(self, relay_state):
+        """Returns a redirection URL from the relay state
+
+        :param relay_state: SAML response's relay state
+        :type relay_state: string
+
+        :return: Redirection URL
+        :rtype: string
+        """
+        relay_state_parse_result = urlparse.urlparse(relay_state)
+        relay_state_parameters = urlparse.parse_qs(relay_state_parse_result.query)
+
+        del relay_state_parameters[self.PROVIDER_NAME]
+        del relay_state_parameters[self.IDP_ENTITY_ID]
+        del relay_state_parameters[self.LIBRARY_SHORT_NAME]
+
+        redirect_uri = urlparse.urlunparse(
+            (
+                relay_state_parse_result.scheme,
+                relay_state_parse_result.netloc,
+                relay_state_parse_result.path,
+                relay_state_parse_result.params,
+                urllib.urlencode(relay_state_parameters),
+                relay_state_parse_result.fragment,
+            )
+        )
+
+        return redirect_uri
+
+    def _redirect_with_error(self, redirect_uri, problem_detail):
+        """Redirects the patron to the given URL, with the given ProblemDetail encoded into the fragment identifier
+
+        :param redirect_uri: Redirection URL
+        :type redirect_uri: string
+
+        :param problem_detail: ProblemDetail object describing the error
+        :type problem_detail: ProblemDetail
+
+        :return: Redirection response
+        :rtype: Response
+        """
+        return redirect(self._error_uri(redirect_uri, problem_detail))
+
+    def saml_authentication_redirect(self, params, db):
+        """Redirects an unauthenticated patron to the authentication URL of the
+        appropriate SAML IdP.
+        Over on that other site, the patron will authenticate and be
+        redirected back to the circulation manager, ending up in
+        saml_authentication_callback.
+
+        :param params: Query parameters
+        :type params: Dict
+
+        :param db: Database session
+        :type db: sqlalchemy.orm.session.Session
+
+        :return: Redirection response
+        :rtype: Response
+        """
+        provider_name = params.get(self.PROVIDER_NAME)
+        idp_entity_id = params.get(self.IDP_ENTITY_ID)
+        redirect_uri = params.get(self.REDIRECT_URI, request.path)
+
+        provider = self._authenticator.saml_provider_lookup(provider_name)
+        if isinstance(provider, ProblemDetail):
+            return self._redirect_with_error(redirect_uri, provider)
+
+        authentication_manager = self._get_authentication_manager(db, provider)
+        relay_state = self._add_params_to_url(
+            redirect_uri,
+            {
+                self.LIBRARY_SHORT_NAME: provider.library(db).short_name,
```

Comment:
In addition to the other issues I pointed out with this, it's not necessary to explicitly pass around a library identifier. We have an annotator, `@has_library`, which modifies a route so that it always operates in the context of a particular library. See oauth_callback in api/routes.py for an example.

Within the controller method, you can refer to flask.request.library to get the Library object. As long as you're using `url_for()` to generate links into the controller, the current library will be propagated into those links.

Reply:
I was thinking about it but, unfortunately, we can't use dynamic routes in the case of SAML callback. It has to be mentioned in the SP metadata which is uploaded to the IdP and it cannot change, otherwise, a SAML request with a SAML callback different from the one configured in the metadata won't pass validation

Reply:
I'm not sure what you mean by dynamic routes here. Is the issue that a given hostname can have one and only one SAML callback? With the library name being "dynamic" information that can go into the query string but not the path portion of the URL?

Reply:
Yes, the URL of the SP's SAML callback has to be constant. 

For example, if the URL of the SP's SAML callback configured as `https://circulation.manager.org/saml_callback`, we can't use `https://circulation.manager.org/LIBRARY_1/saml_callback` for `LIBRARY 1` and `https://circulation.manager.org/LIBRARY_2/saml_callback` for `LIBRARY 2`, we have to use the same URL configured in the IdP's metadata and we have to pass additional parameters as part of the relay state:
1. Inside `SAMLController.saml_authentication_redirect` we add additional parameters to the redirect URL and save it as the relay state.
2. Inside `SAMLController.saml_callback` we extract parameters from the relay state and delete them from the query part of the URL

Reply:
OK, mention this in a comment.